### DISCUSSION TITLE
Apicurio registry base URL

### DISCUSSION THREAD

**Discussion by Bewi42** [Apicurio/apicurio-registry#2134] [Q&A]
Hello,

How do you configure Apicurio so that it knows its base URL when deployed into a Kube pod ?

More details :
Apicurio Registry is deployed on a Kube cluster inside a "schemas-registry" pod along with a "schemas-registry-service" Service that is inside a "namespace1" namespace.
The Kube cluster assigns the service an internal server name that is : "schemas-registry-service.namespace1"
I have an exposition feature that allows me to map the following URL to the service so that it can be accessed from outside the Kube cluster :
https://myexposition/namespace1/schemas-registry-service
BUT when I try to access this URL, Apicurio redirects me to :
http://schemas-registry-service.international-technical-resources/ui/
Instead of :
https://myexposition/namespace1/schemas-registry-service/ui/

Is there a configuration variable to allow Apicurio to know that it should redirect me to the exposed URL instead of the internal one ?

**Comment by Bewi42** [comment]
Documentation points to REGISTRY_UI_CONFIG_APIURL and REGISTRY_UI_CONFIG_UIURL environment variables :
https://www.apicur.io/registry/docs/apicurio-registry/2.1.x/getting-started/assembly-managing-registry-artifacts-ui.html
So I have added these variables to my deployment file (with pointers to external URL) but the redirect still points to the internal URL.
Are these variables taken into account ? Are they translated to a parameter that I could check inside the container ?

**Reply by Bewi42** [reply]
Looks like someone already has had the same problem with no solution :
https://stackoverflow.com/questions/68321783/override-default-base-urls-for-accessing-apicurio-registry-ui

**Reply by EricWittmann** [reply]
Is it the case that you can go directly to the desired URL:  https://myexposition/namespace1/schemas-registry-service/ui/

The problem is just the redirect when you go here?

https://myexposition/namespace1/schemas-registry-service

There has been some work done in this area, and it's weirdly tricky.  Have a look at this:

https://github.com/Apicurio/apicurio-registry/issues/1959

I recommend reading that entire thread for the full context.  But the TLDR is that you can set this ENV variable:

`REGISTRY_ROOT_REDIRECT`

Set it to whatever absolute value you want.  Here is the relevant configuration in `application.properties`:

https://github.com/Apicurio/apicurio-registry/blob/master/app/src/main/resources/application.properties#L48

**Reply by Bewi42** [reply]
Thanks for the answer. However it does not totally work.

I have set the variable like this :
- name: REGISTRY_ROOT_REDIRECT
  value: https://myexposition/namespace1/schemas-registry-service/ui/

It helps with the "/" to "/ui/" redirect, but the static resources (css, js, etc ...) are not loaded because their URL are like this one :
https://myexposition/ui/version.js
Where they should be like :
https://myexposition/namespace1/schemas-registry-service/ui/version.js
And so the UI fails to display.

And, even worse, it does not help at all for the API. When I try this URL :
https://myexposition/namespace1/schemas-registry-service/apis/registry/v2/groups/
I have this message :
{"message":"RESTEASY003210: Could not find resource for full path: http://schemas-registry-service.namespace1/apis/registry/v2/groups/","error_code":404,"detail":"NotFoundException: RESTEASY003210: Could not find resource for full path: http://schemas-registry-service.namespace1/apis/registry/v2/groups/","name":"NotFoundException"}

**Reply by EricWittmann** [reply]
We're probably in some unknown deployment territory for me here.  For example I don't know what an exposition feature is in k8s.  Did you try setting `REGISTRY_PROXY_ADDRESS_FORWARDING` to `true`?

@famartinrh Do you have any insight into this?

**Reply by Bewi42** [reply]
I've tried to set REGISTRY_PROXY_ADDRESS_FORWARDING to true but it does not work either.
When I look to the html response for /ui/ I see that :
<base href="/ui/">
And I think this is why the registry assumes its root context is /ui/ and not /namespace1/schemas-registry-service/ui/

Edit :
Kafka-ui for instance, allows to set a variable to define the root context :
            - name: SERVER_SERVLET_CONTEXT_PATH
              value: /namespace1/myservice

Is there a parameter in apicurio registry that lets you redefine your root context ?

**Reply by EricWittmann** [reply]
OK interesting.  We haven't tested this use-case specifically and so don't have an application config parameter to set the root context.  However, Quarkus might have something helpful built-in.  Have a look at this blog post:

https://quarkus.io/blog/path-resolution-in-quarkus/

Perhaps playing with the `quarkus.http.root-path` or `quarkus.http.non-application-root-path` can help. Apologies that I can't easily try this myself.  

**Note**: Quarkus converts ENV variables to application properties automatically according to Microprofile Config mapping rules:  https://github.com/eclipse/microprofile-config/blob/master/spec/src/main/asciidoc/configsources.asciidoc#environment-variables-mapping-rules

**Comment by Bewi42** [comment]
Hello,

I have tried several things but I can't manage to have a correct base href in the html index document. Each time I get this :
`<base href="/ui/">`
Where I expect to have :
`<base href="/namespace1/schemas-registry-service/ui/">`
I have tried to set following variables :
```
            - name: REGISTRY_ROOT_REDIRECT
              value: https://myexposition/namespace1/schemas-registry-service/ui/
            - name: REGISTRY_UI_CONFIG_UI_CONTEXT_PATH
              value: /namespace1/schemas-registry-service/ui/
            - name: REGISTRY_UI_CONFIG_APIURL
              value: https://myexposition/namespace1/schemas-registry-service/apis/registry/v2
            - name: REGISTRY_UI_CONFIG_UIURL
              value: https://myexposition/namespace1/schemas-registry-service/ui
            - name: QUARKUS_HTTP_ROOT_PATH
              value: /namespace1/schemas-registry-service
            - name: REGISTRY_PROXY_ADDRESS_FORWARDING
              value: "true"
```
I am not sure about the `QUARKUS_HTTP_ROOT_PATH` variable name though. And more important, `quarkus.http.root-path` variable does not appear in `application.properties` file. We have `quarkus.http.non-application-root-path` though, but it is static :
`quarkus.http.non-application-root-path=/`

**Reply by carlesarnal** [reply]
Hi @Bewi42 I'm gonna test this to see if I can come up with a solution to the issue. In the meantime, I want to clarify that Quarkus exposes all the properties present in the application.properties file as environment variables, no matter if we redefine them with a different variable or not.

**Reply by Bewi42** [reply]
Thanks !
